README
This is a "plugin" for the Video Disk Recorder (VDR).

Written by:                  Lars Hanisch <dvb@flensrocker.de>

Project's homepage:          https://github.com/flensrocker/vdr-plugin-dbus2vdr

Latest version available at: git://github.com/flensrocker/vdr-plugin-dbus2vdr.git

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
See the file COPYING for more information.

Description:
This plugin will expose some methods via DBus to control the vdr.

Installation
------------
dbus2vdr will use the system bus so your box has to be configured that the user
in which context your vdr is running is allowed to connect to it.
Copy the de.tvdr.vdr.conf file to /etc/dbus-1/system.d and adjust it to your needs.
It is prepared that only the user 'vdr' can own the used destination and every user
can call its methods.

In the bin-directory you find a helper script for sending dbus messages to the vdr.
Instead of calling:
  dbus-send --system --type=method_call --dest=de.tvdr.vdr \
            --print-reply /Plugins/<pluginname> \
            de.tvdr.vdr.plugin.SVDRPCommand string:'command' string:'parameter'

you can call:
  vdr-dbus-send.sh /Plugins/<pluginname> plugin.SVDRPCommand string:'command' string:'parameter'

Dependencies
------------
Debian/Ubuntu packages:
libdbus-1-dev, libglib2.0-dev, pkg-config, libjpeg-dev, libpng++-dev

Restrictions
------------
Objectpaths and interface names are restricted to some naming conventions. Only letters, numbers
and the underscore are allowed. DBus doesn't provide any official escaping, so dbus2vdr won't do it.
The only part of the names, which could contain other symbols are name of plugins. But "officially"
there are also no other characters than letters and numbers allowed. And I only knew about
streamdev-server/-client which break this rule. I think they should get fixed...

Multi Instance Support
----------------------
The default destination address is "de.tvdr.vdr". In case of a vdr version above 1.7.4 and an
instance id (vdr parameter -i) greater than 0, this instance id is appended to this address,
e.g. "de.tvdr.vdr1". If you use this feature you have to clone the dbus-policy file and change
the address. Also vdr-dbus-send.sh is not ready for this - use dbus-send or whatever you use to
communicate with dbus2vdr...

Commandline Arguments
---------------------
-s, --shutdown-hooks=/path/to/shutdown-hooks
        uses all files in the given directory as shutdown-hooks
        usually it's /usr/share/vdr/shutdown-hooks
-w, --shutdown-hooks-wrapper=/path/to/shutdown-hooks-wrapper
        path to a program that will call the shutdown-hooks with suid
-o, --osd
        creates an OSD-provider which will save the OSD as PNG files
        (see comments below at section "OSD")
--systemd
        use sd_notify to notify systemd
-u, --upstart
        enable Upstart started/stopped events
-n, --network
        enable network support
        (see comments below at section "Network")
--session
        connect to the session D-Bus daemon
--no-system
        don't connect to the system D-Bus daemon
--no-mainloop
        don't start GMainLoop
        (don't use this option if you don't understand)
--nulldevice[=force]
        creates an output device, which does nothing
        can be activated to suspend in/output of DVB streams
        if "force" dbus2vdr will set this device as primary device on startup

Interface "channel"
-------------------
- get count of channels
  vdr-dbus-send.sh /Channels channel.Count

- get current channel
  vdr-dbus-send.sh /Channels channel.Current

- get channels between from_index and to_index
  vdr-dbus-send.sh /Channels channel.GetFromTo int32:from_index int32:to_index

- list channels like SVDRP command CHAN
  vdr-dbus-send.sh /Channels channel.List string:'[ :groups | <number> | <name> | <id> ]'

  The returned array contains the channel number and the string from the channels.conf.

Interface "device"
-------------------
- get current primary device
  vdr-dbus-send.sh /Devices device.GetPrimary

- get index of the nulldevice (-1 if not created)
  vdr-dbus-send.sh /Devices device.GetNullDevice

- request switch to primary device
  vdr-dbus-send.sh /Devices device.RequestPrimary int32:index

- list devices
  vdr-dbus-send.sh /Devices device.List

  The returned array contains the device index, number, a boolean value "hasDecoder",
  a boolean value "isPrimary" and the name of the device.

Interface "epg"
---------------
- disable EIT scanner with timeout (default: 3600)
  vdr-dbus-send.sh /EPG epg.DisableScanner int32:3600

- enable EIT scanner
  vdr-dbus-send.sh /EPG epg.EnableScanner

- clear EPG data with a value for inactivity timeout of eit-scanner (default: 10)
  vdr-dbus-send.sh /EPG epg.ClearEPG string:'channel' int32:10

- add EPG entry
  vdr-dbus-send.sh /EPG epg.PutEntry array:string:'C ...','E ...',...

  The end marker argument '.' will be added if not provided.

- read EPG data from file
  vdr-dbus-send.sh /EPG epg.PutFile string:'/path/to/epg/data'

- get current or next event of given or all channels if string is empty
  vdr-dbus-send.sh /EPG epg.Now string:'channel'
  vdr-dbus-send.sh /EPG epg.Next string:'channel'
  vdr-dbus-send.sh /EPG epg.At string:'channel' uint64:time

  The events are returned as an array of structs.
  Each struct contains a string as a key and a variant as its value.
  Possible keys and their variant type are:
  ChannelID      string
  EventID        uint32
  Title          string
  ShortText      string
  Description    string
  StartTime      uint64
  EndTime        uint64
  Duration       uint64
  Vps            uint64
  RunningStatus  int32
  ParentalRating int32
  ContentID[i]   int32  (0 <= i <= MaxEventContents)
  Content[i]     string (0 <= i <= MaxEventContents)

Interface "plugin"
------------------
If a plugin's name contains a hyphen, it will be replaced with an underscore in its
object-path, e.g. streamdev-server will be found at /Plugins/streamdev_server.

- send SVDRP commands to plugins
  vdr-dbus-send.sh /Plugins/<pluginname> plugin.SVDRPCommand string:'command' string:'parameter'

If the SVDRP command doesn't take a parameter you must pass an empty string.

- call Service method of plugins
  vdr-dbus-send.sh /Plugins/<pluginname> plugin.Service string:'id' string:'data'

Interface "pluginmanager"
------------------
- list all loaded plugins
  vdr-dbus-send.sh /Plugins pluginmanager.List
  You'll get an array of structs containing two strings, the name and the version of the plugins.

Interface "recording"
---------------------
- trigger an update of the list of recordings
  vdr-dbus-send.sh /Recordings recording.Update

- get info about one recording
  vdr-dbus-send.sh /Recordings recording.Get [ variant:int32:number | variant:string:'path' ]

  The following is returned:
  struct of
    int32  number of recording
    array of
      string   key
      variant  value

  possible keys: Path, Name, Title, Start, Deleted, Priority, Lifetime,
  HierarchyLevels, FramesPerSecond, NumFrames, LengthInSeconds, FileSizeMB,
  IsPesRecording, IsNew, IsEdited
  If an cRecordingInfo object exists there can also be:
  Info/ChannelID, Info/ChannelName, Info/Title, Info/ShortText, Info/Description,
  Info/Aux, Info/FramesPerSecond
  If the info object has components there might be:
  Info/Component/<nr>/stream, Info/Component/<nr>/type, Info/Component/<nr>/language,
  Info/Component/<nr>/description

- list recordings
  vdr-dbus-send.sh /Recordings recording.List

  returned is an array of the same structs as with "Get".
  
- play recording given by path or number
  vdr-dbus-send.sh /Recordings recording.Play [ variant:int32:number | variant:string:'path' ] [ variant:string:'hh:mm:ss.f' | variant:int32:framenumber ]
  
  If framenumber is -1, playing is resumed at the last saved position.

If your vdr is patched with the "extra video directories" patch there are also the following methods.

- add extra video directory
  vdr-dbus-send.sh /Recordings recording.AddExtraVideoDirectory string:directory

  Don't forget to manually update your recordings after adding all directories.

- clear extra video directories list
  vdr-dbus-send.sh /Recordings recording.ClearExtraVideoDirectories

- delete an extra video directory
  vdr-dbus-send.sh /Recordings recording.DeleteExtraVideoDirectory string:directory

- list all extra video directories
  vdr-dbus-send.sh /Recordings recording.ListExtraVideoDirectories

  Besides the usual reply-code and -message you get an array of strings with the
  currently present video directories.

Interface "remote"
------------------
- enable/disable the remote
  vdr-dbus-send.sh /Remote remote.Enable
  vdr-dbus-send.sh /Remote remote.Disable
  vdr-dbus-send.sh /Remote remote.Status

- hit a key, you may specify more than one key
  vdr-dbus-send.sh /Remote remote.HitKey string:'Menu'
  vdr-dbus-send.sh /Remote remote.HitKeys array:string:'Menu','Down','Down','Ok',...

- display list of strings on the osd and let the user select one

HISTORY
VDR Plugin 'dbus2vdr' Revision History
--------------------------------------

2011-03-20: Version 0.0.1

- Initial revision.

2011-03-21: Version 0.0.1a

- use thread pool for handling dbus messages so multiple
  message won't block each other

2011-03-21: Version 0.0.1b

- refactoring of message dispatcher to make it easier to add
  new interfaces

2011-03-22: Version 0.0.1c

- add call to Service method of plugins

2011-03-22: Version 0.0.2

- add new interface for reading epg-data from a file

2011-03-22: Version 0.0.2a

- use system bus, not session bus

2011-03-29: Version 0.0.2b

- add new interface for getting infos about the next timer
- rework monitor loop to work with introspection (needed for
  python binding)

2011-04-20: Version 0.0.2e

- add new interface for the remote (enable, disable, hitkey)

2011-04-27: Version 0.0.2f

- add CallPlugin to remote interface

2011-04-27: Version 0.0.2g

- add new recording interface, for now only an update of the
  recording list can be triggered

2011-05-13: Version 0.0.2h

- add new skin interface with "QueueMessage" to display a message
  on the OSD (thanks to Holger Schvestka)

2011-05-13: Version 0.0.2i

- add ClearEPG to EPG interface

2011-06-08:

- use cEITFilter::SetDisableUntil only if vdr >= 1.7.11

2011-09-22: Version 0.0.2j

- process NameAcquired signal auf DBus

2011-12-10: Version 0.0.3

- add new shutdown interface with "ConfirmShutdown"
  returns a reply code of 250 if the user is inactive
  and vdr has nothing to do in the near future

2011-12-11: Version 0.0.3a

- make "ConfirmShutdown" a bit more talkative

2011-12-14: Version 0.0.3b

- don't call vdr's ConfirmShutdown if remote is disabled
  that would always return false

2011-12-15: Version 0.0.3c

- add commandline argument: --shutdown-hooks=/path/to/vdr/shutdown-hooks

2012-01-07: Version 0.0.3d

- add PutEntry to EPG interface

2012-01-18: Version 0.0.3e

- epg.c: correct introspection data for PutEntry
- Makefile: add WANT_I18N (thanks to Christian Ruppert)
- Makefile: don't override "directory environment" (thanks to Christian Ruppert)

2012-01-26: Version 0.0.3f

- add new setup interface with "Set" and "Get" for selected parameters

2012-01-26: Version 0.0.3g

- add "List" to setup interface

2012-02-08: Version 0.0.3h

- add "AskUser" to remote interface

2012-02-19: Version 0.0.3i

- shutdown-wrapper is a wrapper which tests the shutdown-hooks
  set its SUID bit if needed

2012-02-23: Version 0.0.3j

- add "Now" and "Next" to EPG interface

2012-03-07: Version 0.0.3k

- add "DisableScanner" and "EnableScanner" to EPG interface

2012-03-08: Version 0.0.4

- add OSD interface

2012-03-10: Version 0.0.4a

- add "CreateProvider" and "DeleteProvider" to OSD interface

2012-03-11: Version 0.0.4b

- adapt to vdr 1.7.26 and add missing symbol

2012-04-04: Version 0.0.4c

- accept multiple keys with "HitKey"

2012-04-12: Version 0.0.4d

- add "At" to EPG interface

2012-04-19: Version 0.0.5
- add "CurrentSkin" and "ListSkins" to Skin interface

2012-04-26: Version 0.0.6
- send upstart signals for every plugin on startup and stopping

2012-04-26: Version 0.0.6a
- use a DBus method call for Upstart, not a DBus signal

2012-04-28: Version 0.0.6b
- emit only one Upstart event for all plugins

2012-05-02: Version 0.0.6c
- emit Upstart signal only if activated and send a SIGSTOP

2012-05-13: Version 0.0.6d
- if dbus disconnects try to reconnect to system bus

2012-05-15: Version 0.0.6e
- add support for getting/setting plugin parameters in the setup.conf

2012-05-23: Version 0.0.6f
- correct parameter parsing of epg.PutEntry and remote.AskUser

2012-05-23: Version 0.0.6g
- add "Del" to setup interface

2012-05-23: Version 0.0.7
- output all plugin settings with setup.List
- delete all plugin settings with "Del plugin.*"

2012-06-06: Version 0.0.7a
- extend "ConfirmShutdown" with an "ignore user" parameter

2012-06-14: Version 0.0.7b
- refactor connection handling
- waiting for "NameAcquired" before sending signals

2012-06-20: Version 0.0.7c
- call dbus_threads_init_default, since connection shall be used by multiple threads

2012-06-27: Version 0.0.7d
- add NextWakeupTime to Setup interface
- add ManualStart to Shutdown interface

2012-07-04: Version 0.0.7e
- fix value range test in setup.Set for int32
- add support for getting all values from setup.conf
- add support for storing arbitrary values in setup.conf

2012-07-05: Version 0.0.7f
- fix a memory leak in setup.Set
- fix a bug while setting unsupported values in setup.Set

2012-07-12: Version 0.0.8
- fix the dbus main loop

2012-07-24: Version 0.0.8a

epg.c
#include "epg.h"
#include "common.h"
#include "helper.h"

#include <limits.h>

#include <vdr/eit.h>
#include <vdr/epg.h>
#include <vdr/svdrp.h>
#include <vdr/timers.h>


namespace cDBusEpgHelper
{
  enum eMode   { dmmAll, dmmPresent, dmmFollowing, dmmAtTime };

  static const char *_xmlNodeInfo = 
    "<!DOCTYPE node PUBLIC \"-//freedesktop//DTD D-BUS Object Introspection 1.0//EN\"\n"
    "       \"http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd\">\n"
    "<node>\n"
    "  <interface name=\""DBUS_VDR_EPG_INTERFACE"\">\n"
    "    <method name=\"DisableScanner\">\n"
    "      <arg name=\"eitdisabletime\" type=\"i\" direction=\"in\"/>\n"
    "      <arg name=\"replycode\"      type=\"i\" direction=\"out\"/>\n"
    "      <arg name=\"replymessage\"   type=\"s\" direction=\"out\"/>\n"
    "    </method>\n"
    "    <method name=\"EnableScanner\">\n"
    "      <arg name=\"replycode\"      type=\"i\" direction=\"out\"/>\n"
    "      <arg name=\"replymessage\"   type=\"s\" direction=\"out\"/>\n"
    "    </method>\n"
    "    <method name=\"ClearEPG\">\n"
    "      <arg name=\"channel\"        type=\"s\" direction=\"in\"/>\n"
    "      <arg name=\"eitdisabletime\" type=\"i\" direction=\"in\"/>\n"
    "      <arg name=\"replycode\"      type=\"i\" direction=\"out\"/>\n"
    "      <arg name=\"replymessage\"   type=\"s\" direction=\"out\"/>\n"
    "    </method>\n"
    "    <method name=\"PutEntry\">\n"
    "      <arg name=\"entryline\"      type=\"as\" direction=\"in\"/>\n"
    "      <arg name=\"replycode\"      type=\"i\"  direction=\"out\"/>\n"
    "      <arg name=\"replymessage\"   type=\"s\"  direction=\"out\"/>\n"
    "    </method>\n"
    "    <method name=\"PutFile\">\n"
    "      <arg name=\"filename\"       type=\"s\" direction=\"in\"/>\n"
    "      <arg name=\"replycode\"      type=\"i\" direction=\"out\"/>\n"
    "      <arg name=\"replymessage\"   type=\"s\" direction=\"out\"/>\n"
    "    </method>\n"
    "    <method name=\"Now\">\n"
    "      <arg name=\"channel\"        type=\"s\"  direction=\"in\"/>\n"
    "      <arg name=\"replycode\"      type=\"i\"  direction=\"out\"/>\n"
    "      <arg name=\"replymessage\"   type=\"s\"  direction=\"out\"/>\n"
    "      <arg name=\"event_list\"     type=\"aa(sv)\" direction=\"out\"/>\n"
    "    </method>\n"
    "    <method name=\"Next\">\n"
    "      <arg name=\"channel\"        type=\"s\"  direction=\"in\"/>\n"
    "      <arg name=\"replycode\"      type=\"i\"  direction=\"out\"/>\n"
    "      <arg name=\"replymessage\"   type=\"s\"  direction=\"out\"/>\n"
    "      <arg name=\"event_list\"     type=\"aa(sv)\" direction=\"out\"/>\n"
    "    </method>\n"
    "    <method name=\"At\">\n"
    "      <arg name=\"channel\"        type=\"s\"  direction=\"in\"/>\n"
    "      <arg name=\"time\"           type=\"t\"  direction=\"in\"/>\n"
    "      <arg name=\"replycode\"      type=\"i\"  direction=\"out\"/>\n"
    "      <arg name=\"replymessage\"   type=\"s\"  direction=\"out\"/>\n"
    "      <arg name=\"event_list\"     type=\"aa(sv)\" direction=\"out\"/>\n"
    "    </method>\n"
    "  </interface>\n"
    "</node>\n";

  static void DisableScanner(cDBusObject *Object, GVariant *Parameters, GDBusMethodInvocation *Invocation)
  {
  #if APIVERSNUM >= 10711
    int eitDisableTime = 3600; // one hour should be ok as a default
    int s = 0;
    g_variant_get(Parameters, "(i)", &s);
    if (s > 0)
       eitDisableTime = s;
    cString replyMessage = cString::sprintf("EIT scanner disabled for %d sseconds", eitDisableTime);
    isyslog("dbus2vdr: %s.DisableScanner: %s", DBUS_VDR_EPG_INTERFACE, *replyMessage);
    cEitFilter::SetDisableUntil(time(NULL) + eitDisableTime);
    cDBusHelper::SendReply(Invocation, 250, *replyMessage);
  #else
    esyslog("dbus2vdr: %s.DisableScanner: you need at least vdr 1.7.11", DBUS_VDR_EPG_INTERFACE);
    cDBusHelper::SendReply(Invocation, 550, "you need at least vdr 1.7.11");
  #endif
  };

  static void EnableScanner(cDBusObject *Object, GVariant *Parameters, GDBusMethodInvocation *Invocation)
  {
  #if APIVERSNUM >= 10711
    isyslog("dbus2vdr: %s.EnableScanner: EIT scanner enabled", DBUS_VDR_EPG_INTERFACE);
    cEitFilter::SetDisableUntil(0);
    cDBusHelper::SendReply(Invocation, 250, "EIT scanner enabled");
  #else
    esyslog("dbus2vdr: %s.EnableScanner: you need at least vdr 1.7.11", DBUS_VDR_EPG_INTERFACE);
    cDBusHelper::SendReply(Invocation, 550, "you need at least vdr 1.7.11");
  #endif
  };

  static void ClearEPG(cDBusObject *Object, GVariant *Parameters, GDBusMethodInvocation *Invocation)
  {
    const char *channel = NULL;
    int timeout = 0;
    g_variant_get(Parameters, "(&si)", &channel, &timeout);
    int eitDisableTime = 10; // seconds until EIT processing is enabled again after a CLRE command
    if (timeout > 0) {
       eitDisableTime = timeout;
       isyslog("dbus2vdr: %s.ClearEPG: using %d seconds as EIT disable timeout", DBUS_VDR_EPG_INTERFACE, eitDisableTime);
       }

    if (channel) {
       tChannelID ChannelID = tChannelID::InvalidID;
       if (isnumber(channel)) {
          int o = strtol(channel, NULL, 10);
#if VDRVERSNUM > 20300
          LOCK_CHANNELS_READ;
          const cChannels *channels = Channels;
#else
          cChannels *channels = &Channels;
#endif
          if (o >= 1 && o <= channels->MaxNumber())
             ChannelID = channels->GetByNumber(o)->GetChannelID();
          }
       else {
          ChannelID = tChannelID::FromString(channel);
          if (ChannelID == tChannelID::InvalidID) {
             const cChannels *channels = NULL;
#if VDRVERSNUM > 20300
             LOCK_CHANNELS_READ;
             channels = Channels;
#else
             channels = &Channels;
#endif
             for (const cChannel *Channel = channels->First(); Channel; Channel = channels->Next(Channel)) {
                 if (!Channel->GroupSep()) {
                    if (strcasecmp(Channel->Name(), channel) == 0) {
                       ChannelID = Channel->GetChannelID();
                       break;
                       }
                    }
                 }
             }
          }
       if (!(ChannelID == tChannelID::InvalidID)) {
#if VDRVERSNUM > 20300
          cStateKey StateKey;
          cSchedules *s = cSchedules::GetSchedulesWrite(StateKey, 1000);
#else
          cSchedulesLock SchedulesLock(true, 1000);
          cSchedules *s = (cSchedules *)cSchedules::Schedules(SchedulesLock);
#endif
          if (s) {
             cSchedule *Schedule = NULL;
             ChannelID.ClrRid();
             for (cSchedule *p = s->First(); p; p = s->Next(p)) {
                 if (p->ChannelID() == ChannelID) {
                    Schedule = p;
                    break;
                    }
                 }
             if (Schedule) {
                Schedule->Cleanup(INT_MAX);
                #if APIVERSNUM >= 10711
                cEitFilter::SetDisableUntil(time(NULL) + eitDisableTime);
                #endif
                cString replyMessage = cString::sprintf("EPG data of channel \"%s\" cleared", channel);
                cDBusHelper::SendReply(Invocation, 250, *replyMessage);
                }
             else {
                cString replyMessage = cString::sprintf("No EPG data found for channel \"%s\"", channel);
                cDBusHelper::SendReply(Invocation, 550, *replyMessage);
                }
#if VDRVERSNUM > 20300
             StateKey.Remove();
#endif
             }
          else
             cDBusHelper::SendReply(Invocation, 451, "Can't get EPG data");
          }
       else {
          cString replyMessage = cString::sprintf("Undefined channel \"%s\"", channel);
          cDBusHelper::SendReply(Invocation, 501, *replyMessage);
          }
       }
    else {
#if VDRVERSNUM > 20300
       LOCK_TIMERS_WRITE;
       LOCK_SCHEDULES_WRITE;
       for (cTimer *Timer = Timers->First(); Timer; Timer = Timers->Next(Timer))
           Timer->SetEvent(NULL); // processing all timers here (local *and* remote)
       for (cSchedule *Schedule = Schedules->First(); Schedule; Schedule = Schedules->Next(Schedule))
           Schedule->Cleanup(INT_MAX);
#else
       cSchedules::ClearAll();
#endif
       cEitFilter::SetDisableUntil(time(NULL) + eitDisableTime);
       cDBusHelper::SendReply(Invocation, 250, "EPG data cleared");
       }
  };

#if VDRVERSNUM > 20300
// copied from svdrp.c, maybe it will be available in future versions?
// --- cPUTEhandler ----------------------------------------------------------

class cPUTEhandler {
private:
  FILE *f;
  int status;
  const char *message;
public:
  cPUTEhandler(void);
  ~cPUTEhandler();
  bool Process(const char *s);
  int Status(void) { return status; }
  const char *Message(void) { return message; }
  };

cPUTEhandler::cPUTEhandler(void)
{
  if ((f = tmpfile()) != NULL) {
     status = 354;
     message = "Enter EPG data, end with \".\" on a line by itself";
     }
  else {
     LOG_ERROR;
     status = 554;
     message = "Error while opening temporary file";
     }
}

cPUTEhandler::~cPUTEhandler()
{
  if (f)
     fclose(f);
}

bool cPUTEhandler::Process(const char *s)
{
  if (f) {
     if (strcmp(s, ".") != 0) {
        fputs(s, f);
        fputc('\n', f);
        return true;
        }
     else {
        rewind(f);
        if (cSchedules::Read(f)) {
           cSchedules::Cleanup(true);
           status = 250;
           message = "EPG data processed";
           }
        else {
           status = 451;
           message = "Error while processing EPG data";
           }
        fclose(f);
        f = NULL;
        }
     }
  return false;
}
#endif

  static void PutEntry(cDBusObject *Object, GVariant *Parameters, GDBusMethodInvocation *Invocation)
  {
    gsize len = 0;
    GVariant *array = g_variant_get_child_value(Parameters, 0);
    const gchar **line = g_variant_get_strv(array, &len);
    if (len < 1) {
       g_free(line);
       cDBusHelper::SendReply(Invocation, 501, "at least one element must be given");
       return;
       }

    cPUTEhandler *handler = new cPUTEhandler();
    if (handler->Status() == 354) {
       for (gsize i = 0; i < len; i++) {
           d4syslog("dbus2vdr: %s.PutEntry: item = %s", DBUS_VDR_EPG_INTERFACE, line[i]);
           d4syslog("dbus2vdr: %s.PutEntry: status = %d, message = %s", DBUS_VDR_EPG_INTERFACE, handler->Status(), handler->Message());
           if (!handler->Process(line[i]))
              break;
           }
       d4syslog("dbus2vdr: %s.PutEntry: status = %d, message = %s", DBUS_VDR_EPG_INTERFACE, handler->Status(), handler->Message());
       if (handler->Status() == 354)
          handler->Process(".");
       }
    cDBusHelper::SendReply(Invocation, handler->Status(), handler->Message());
    delete handler;
    g_free(line);
  };

  static void PutFile(cDBusObject *Object, GVariant *Parameters, GDBusMethodInvocation *Invocation)
  {
    const char *filename = NULL;
    g_variant_get(Parameters, "(&s)", &filename);

    if ((filename != NULL) && (*filename != 0)) {
       FILE *f = fopen(filename, "r");
       if (f) {
          cString message = cString::sprintf("start reading epg data from %s", filename);
          cDBusHelper::SendReply(Invocation, 0, *message);
          if (cSchedules::Read(f))
             cSchedules::Cleanup(true);
          fclose(f);
          }
       else {
          esyslog("dbus2vdr: %s.PutFile: error opening %s", DBUS_VDR_EPG_INTERFACE, filename);
          cString message = cString::sprintf("error opening %s", filename);
          cDBusHelper::SendReply(Invocation, -2, *message);
          }
       }
    else
       cDBusHelper::SendReply(Invocation, -1, "no filename");
  };

  static void sAddEvent(GVariantBuilder *Array, const cEvent &Event)
  {
    const char *c;
    guint32 tu32;
    guint64 tu64;
    int ti;
    gboolean tb;

    GVariantBuilder *arr = g_variant_builder_new(G_VARIANT_TYPE("a(sv)"));

    cString cid = Event.ChannelID().ToString();
    c = *cid;
    cDBusHelper::AddKeyValue(arr, "ChannelID", "s", (void**)&c);

    tu32 = Event.EventID();
    cDBusHelper::AddKeyValue(arr, "EventID", "u", (void**)&tu32);

    c = Event.Title();
    if (c != NULL)
       cDBusHelper::AddKeyValue(arr, "Title", "s", (void**)&c);

    c = Event.ShortText();
    if (c != NULL)
       cDBusHelper::AddKeyValue(arr, "ShortText", "s", (void**)&c);

    c = Event.Description();
    if (c != NULL)
       cDBusHelper::AddKeyValue(arr, "Description", "s", (void**)&c);

    tu64 = Event.StartTime();
    cDBusHelper::AddKeyValue(arr, "StartTime", "t", (void**)&tu64);

    tu64 = Event.EndTime();
    cDBusHelper::AddKeyValue(arr, "EndTime", "t", (void**)&tu64);

    tu64 = Event.Duration();
    cDBusHelper::AddKeyValue(arr, "Duration", "t", (void**)&tu64);

    tu64 = Event.Vps();
    cDBusHelper::AddKeyValue(arr, "Vps", "t", (void**)&tu64);

    ti = Event.RunningStatus();
    cDBusHelper::AddKeyValue(arr, "RunningStatus", "i", (void**)&ti);

  #if VDRVERSNUM >= 10711
    ti = Event.ParentalRating();
    cDBusHelper::AddKeyValue(arr, "ParentalRating", "i", (void**)&ti);

    tb = Event.HasTimer();
    cDBusHelper::AddKeyValue(arr, "HasTimer", "b", (void**)&tb);

    for (int i = 0; i < MaxEventContents; i++) {
        tu32 = Event.Contents(i);
        if (tu32 != 0) {
           cDBusHelper::AddKeyValue(arr, *cString::sprintf("ContentID[%d]", i), "u", (void**)&tu32);
           c = cEvent::ContentToString(tu32);
           cDBusHelper::AddKeyValue(arr, *cString::sprintf("Content[%d]", i), "s", (void**)&c);
           }
        }
  #endif

    g_variant_builder_add_value(Array, g_variant_builder_end(arr));
    g_variant_builder_unref(arr);
  }

#if VDRVERSNUM > 20300
  static bool sGetChannel(GVariant *Arg, const char **Input, const cChannels* Channels, const cChannel **Channel)
#else
  static bool sGetChannel(GVariant *Arg, const char **Input, cChannels* Channels, const cChannel **Channel)
#endif
  {
    *Channel = NULL;
    *Input = NULL;
    if (g_variant_is_of_type(Arg, G_VARIANT_TYPE_STRING)) {
       g_variant_get(Arg, "&s", Input);
       if (**Input == 0)
          return true;
       if (isnumber(*Input))
          *Channel = Channels->GetByNumber(strtol(*Input, NULL, 10));
       else
          *Channel = Channels->GetByChannelID(tChannelID::FromString(*Input));
       if (*Channel == NULL)
          return false;
       }
    return true;
  }

  static void sReturnError(GDBusMethodInvocation *Invocation, int  ReplyCode, const char *ReplyMessage)
  {
    GVariantBuilder *builder = g_variant_builder_new(G_VARIANT_TYPE("(isaa(sv))"));
    g_variant_builder_add(builder, "i", ReplyCode);
    g_variant_builder_add(builder, "s", ReplyMessage);
    GVariantBuilder *array = g_variant_builder_new(G_VARIANT_TYPE("aa(sv)"));
    GVariantBuilder *innerArray = g_variant_builder_new(G_VARIANT_TYPE("a(sv)"));
    g_variant_builder_add_value(array, g_variant_builder_end(innerArray));
    g_variant_builder_add_value(builder, g_variant_builder_end(array));
    g_dbus_method_invocation_return_value(Invocation, g_variant_builder_end(builder));
    g_variant_builder_unref(innerArray);
    g_variant_builder_unref(array);
    g_variant_builder_unref(builder);
  };
  
  static void sGetEntries(cDBusObject *Object, GVariant *Parameters, GDBusMethodInvocation *Invocation, eMode mode)
  {
#if VDRVERSNUM > 20300
    LOCK_CHANNELS_READ;
    const cChannels *channels = Channels;
#else
    cChannels *channels = &Channels;
#endif
    const cChannel *channel = NULL;
    guint64 atTime = 0;
    GVariant *first = g_variant_get_child_value(Parameters, 0);
    
    const char *c = NULL;
    if (!sGetChannel(first, &c, channels, &channel)) {
       cString reply = cString::sprintf("channel \"%s\" not defined", c);
       esyslog("dbus2vdr: %s.GetEntries: %s", DBUS_VDR_EPG_INTERFACE, *reply);
       sReturnError(Invocation, 501, *reply);
       g_variant_unref(first);
       return;
       }
    if (mode == dmmAtTime) {
       GVariant *second = g_variant_get_child_value(Parameters, 1);
       g_variant_get(second, "t", &atTime);
       g_variant_unref(second);
       if (atTime == 0) {
          cString reply = cString::sprintf("missing time");
          esyslog("dbus2vdr: %s.GetEntries: %s", DBUS_VDR_EPG_INTERFACE, *reply);
          sReturnError(Invocation, 501, *reply);
          g_variant_unref(first);
          return;
          }
       }
    g_variant_unref(first);

    const cSchedules *scheds = NULL;
#if VDRVERSNUM > 20300
    cStateKey StateKey;
    scheds = cSchedules::GetSchedulesRead(StateKey, 1000);
#else
    cSchedulesLock sl(false, 1000);
    if (!sl.Locked()) {
       sReturnError(Invocation, 550, "got no lock on schedules");
       return;
       }

    scheds = cSchedules::Schedules(sl);
#endif
    if (scheds == NULL) {
       sReturnError(Invocation, 550, "got no schedules");
       return;
       }

    GVariantBuilder *builder = g_variant_builder_new(G_VARIANT_TYPE("(isaa(sv))"));
    g_variant_builder_add(builder, "i", 250);
    g_variant_builder_add(builder, "s", "");
    GVariantBuilder *array = g_variant_builder_new(G_VARIANT_TYPE("aa(sv)"));

    bool next = false;
    if (channel == NULL) {
       channel = channels->First();
       next = true;
       }

    while (channel) {
          const cSchedule *s = scheds->GetSchedule(channel, false);
          if (s != NULL) {
             const cEvent *e = NULL;
             switch (mode) {
               case dmmPresent:
                 e = s->GetPresentEvent();
                 break;
               case dmmFollowing:
                 e = s->GetFollowingEvent();
                 break;
               case dmmAtTime:
                 e = s->GetEventAround(atTime);
                 break;
               default:
                 e = NULL;
                 break;
               }
             if (e != NULL)
                sAddEvent(array, *e);
             }
          if (next)
             channel = channels->Next(channel);
          else
             break;
          }

#if VDRVERSNUM > 20300
    StateKey.Remove();
#endif
    g_variant_builder_add_value(builder, g_variant_builder_end(array));
    g_dbus_method_invocation_return_value(Invocation, g_variant_builder_end(builder));
    g_variant_builder_unref(array);
    g_variant_builder_unref(builder);
  };

  static void Now(cDBusObject *Object, GVariant *Parameters, GDBusMethodInvocation *Invocation)
  {
    sGetEntries(Object, Parameters, Invocation, dmmPresent);
  };

  static void Next(cDBusObject *Object, GVariant *Parameters, GDBusMethodInvocation *Invocation)
  {
    sGetEntries(Object, Parameters, Invocation, dmmFollowing);
  };

  static void At(cDBusObject *Object, GVariant *Parameters, GDBusMethodInvocation *Invocation)
  {
    sGetEntries(Object, Parameters, Invocation, dmmAtTime);
  };
}

cDBusEpg::cDBusEpg(void)
:cDBusObject("/EPG", cDBusEpgHelper::_xmlNodeInfo)
{
  AddMethod("DisableScanner", cDBusEpgHelper::DisableScanner);
  AddMethod("EnableScanner", cDBusEpgHelper::EnableScanner);
  AddMethod("ClearEPG", cDBusEpgHelper::ClearEPG);
  AddMethod("PutEntry", cDBusEpgHelper::PutEntry);
  AddMethod("PutFile", cDBusEpgHelper::PutFile);
  AddMethod("Now", cDBusEpgHelper::Now);
  AddMethod("Next", cDBusEpgHelper::Next);
  AddMethod("At", cDBusEpgHelper::At);
}

cDBusEpg::~cDBusEpg(void)
{
}
